authz/src/main/java/io/grpc/authz/AuthorizationPolicyTranslator.java

```diff
@@ -0,0 +1,197 @@
+import io.envoyproxy.envoy.config.rbac.v3.Permission;
+import io.envoyproxy.envoy.config.rbac.v3.Policy;
+import io.envoyproxy.envoy.config.rbac.v3.Principal;
+import io.envoyproxy.envoy.config.rbac.v3.Principal.Authenticated;
+import io.envoyproxy.envoy.config.rbac.v3.RBAC;
+import io.envoyproxy.envoy.config.rbac.v3.RBAC.Action;
+import io.envoyproxy.envoy.config.route.v3.HeaderMatcher;
+import io.envoyproxy.envoy.type.matcher.v3.PathMatcher;
+import io.envoyproxy.envoy.type.matcher.v3.StringMatcher;
+import io.grpc.internal.JsonParser;
+import io.grpc.internal.JsonUtil;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * Translates a gRPC SDK authorization policy in JSON string to Envoy RBAC policies.
+ */
+public class AuthorizationPolicyTranslator {
+  private static final ImmutableList<String> UNSUPPORTED_HEADERS = ImmutableList.of(
+      "host", "connection", "keep-alive", "proxy-authenticate", "proxy-authorization",
+      "te", "trailer", "transfer-encoding", "upgrade");
+
+  private static StringMatcher getStringMatcher(String value) {
+    if (value.equals("*")) {
+      return StringMatcher.newBuilder().setPrefix("").build();
+    } else if (value.startsWith("*")) {
+      return StringMatcher.newBuilder().setSuffix(value.substring(1)).build();
+    } else if (value.endsWith("*")) {
+      return StringMatcher.newBuilder().setPrefix(value.substring(0, value.length() - 1)).build();
+    }
+    return StringMatcher.newBuilder().setExact(value).build();
+  }
+
+  private static Principal parseSource(Map<String, ?> source) {
+    List<String> principalsList = JsonUtil.getListOfStrings(source, "principals");
+    if (principalsList == null) {
+      return Principal.newBuilder().setAny(true).build();
+    }
+    if (principalsList.isEmpty()) {
+      return Principal.newBuilder().setAny(true).build();
+    }
+    Principal.Set.Builder principalsSet = Principal.Set.newBuilder();
+    for (String principal: principalsList) {           
+      principalsSet.addIds(
+          Principal.newBuilder().setAuthenticated(
+            Authenticated.newBuilder().setPrincipalName(
+              getStringMatcher(principal)).build()).build());
+    }
+    return Principal.newBuilder().setOrIds(principalsSet.build()).build();
+  }
+
+  private static Permission parseHeader(Map<String, ?> header) throws IllegalArgumentException {
+    String key = JsonUtil.getString(header, "key");
+    if (key == null) {
+      throw new IllegalArgumentException("\"key\" is absent.");
+    }
+    if (key.charAt(0) == ':'
+        || key.startsWith("grpc-")
+        || UNSUPPORTED_HEADERS.contains(key.toLowerCase())) {
+      throw new IllegalArgumentException(String.format("Unsupported \"key\" %s.", key));
+    }
+    List<String> valuesList = JsonUtil.getListOfStrings(header, "values");
+    if (valuesList == null || (valuesList != null && valuesList.isEmpty())) {
```

Comment:
You can depend on the short-circuiting. `valuesList == null || valuesList.isEmpty()`

Reply:
Done